File: cases/taint/csharp/recursive-carry-positive/RecursiveCarryPositive.cs
namespace DataFlowBench;

static class RecursiveCarryPositive
{
    static string dfb_source() // DFB-SOURCE: recursive-carry-input
    {
        return "tainted";
    }

    static void dfb_sink(string value) { } // DFB-SINK: recursive-carry-sink

    static string Carry(string value, int depth)
    {
        if (depth == 0)
        {
            return value; // DFB-WITNESS: recursive-carry-base
        }

        return Carry(value, depth - 1);
    }

    static void Run()
    {
        dfb_sink(Carry(dfb_source(), 5));
    }
}

File: cases/taint/cpp/recursive-carry-positive/recursive_carry_positive.cpp
#include <string>

std::string dfb_source() { // DFB-SOURCE: recursive-carry-input
    return "tainted";
}

void dfb_sink(const std::string &value) {} // DFB-SINK: recursive-carry-sink

std::string carry(const std::string &value, int depth) {
    if (depth == 0) {
        return value; // DFB-WITNESS: recursive-carry-base
    }
    return carry(value, depth - 1);
}

void run() {
    dfb_sink(carry(dfb_source(), 5));
}
